### PR TITLE
Store addressbook

### PR DESCRIPTION
**Pull request by elaineo**
Issue #17

### CODE REVIEW

File: src/rpc/mod.rs [hunk 1]
```diff
@@ -289,6 +290,61 @@ pub fn start(addr: &SocketAddr, client_addr: &SocketAddr, base_path: Option<Path
...
+        let addressbook = addressbook.clone();
+
+        io.add_async_method("emerald_addressBook",
+                            move |_| futures::finished(Value::Array(addressbook.list())).boxed());
+    }
+
+    {
+        let addressbook = addressbook.clone();
+
+        io.add_async_method("emerald_addAddress", move |p| match p {
```
Comment (dulanov) [inline src/rpc/mod.rs]: Three times duplicated code, i don't know exactly, but probably there is a way to collapse them.

File: tests/addressbook_test.rs [hunk 2]
```diff
@@ -0,0 +1,52 @@
+extern crate emerald;
+
+extern crate serde;
+extern crate serde_json;
+extern crate tempdir;
+
+use self::serde_json::Value;
```
Comment (dulanov) [inline tests/addressbook_test.rs]: Without `self` better, just `use serde_json::Value`

File: src/addressbook.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    InvalidAddress,
+}
+
+impl Addressbook {
+    /// Initialize new addressbook service for a dir
+    pub fn new(dir: PathBuf) -> Addressbook {
+        Addressbook { dir: dir }
+    }
+
+    fn read_json(path: &Path) -> Result<serde_json::Value, AddressbookError> {
```
Comment (dulanov) [inline src/addressbook.rs]: Our prefer way, for example, ` pub fn search_by_address<P: AsRef<Path>>(addr: &Address, path: P) -> Result<KeyFile, Error>`.

File: src/addressbook.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        match File::open(path) {
+            Ok(f) => serde_json::from_reader(f).or(Err(AddressbookError::IO)),
+            Err(_) => Err(AddressbookError::IO),
+        }
+    }
+
+    /// List all entries in the addressbook
+    pub fn list(&self) -> Vec<serde_json::Value> {
+        let files = glob(&format!("{}/*.json", &self.dir.to_str().unwrap())).unwrap();
```
Comment (dulanov) [inline src/addressbook.rs]: In production better to use `expect` method instead of `unwrap`, clippy rust linter show a warning for every `unwrap` call.

**Comment by dulanov** [comment]
@elaineo what do you think about Rust, first  impressions?

File: src/addressbook.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+        let mut filename: PathBuf = self.dir.clone();
+        filename.push(format!("{}.json", addr));
+        let mut f = File::create(filename.as_path()).unwrap();
+        match serde_json::to_writer_pretty(&mut f, entry) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(AddressbookError::IO),
+        }
+    }
+
+    /// Edit address entry in addressbook storage (address cannot change)
```
Comment (whilei) [inline src/addressbook.rs]: typo: should be Delete

File: src/rpc/mod.rs (hunk 1 above)
Comment (elaineo) [inline src/rpc/mod.rs]: Yeah, it is not optimal. I think the methods will contain more functionality in the future so I will leave them separate for now.

**Comment by elaineo** [comment]
@dulanov I don't think I understand it well enough to appreciate the benefits yet, but I really want to learn more about Rust :)

**Pull request merged**